meteor-bundle-main.js

```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
           // hasn't given us a signed certificate yet, the response-json
           // won't exist.
           if ((e.code === 'EACCES') || (e.code == 'ENOENT')) {
-            console.log("Skipping unreadable HTTPS key information file:", metadataFilename);
+            console.log("Skipping unreadable HTTPS key information file while examining:",
+                        metadataFilename);
             continue;
           } else {
-            throw e;
+            // Sometimes the server gives us a 0-byte response,
+            // presumably due to a connection getting reset?.
+            console.error("Got exception reading JSON from file:", metadataFilename, "Dazed and confused, but trying to continue.", e);
+            continue;
           }
         }
 
-        // If this certificate isn't valid yet, keep looping, hoping
-        // to find one that is valid.
-        var notBefore = metadata['notBefore'];
-        if (notBefore && (notBefore < nowUnixTimestamp)) {
+        // If the cert is expired, definitely don't use it.
+        if (now > metadata.notAfter) {
+          console.log("Skipping", keyFilename, "because", metadata.notAfter, "is in the past.");
           continue;
         }
 
-        result['ca'] = metadata['ca'];
-        result['key'] = fs.readFileSync(keyFilename, 'utf-8');
-        result['cert'] = metadata['cert'];
-        return result;
+        validCertificates.push(metadata);
+      }
+
+      // Sort by notBefore, ascending.
+      validCertificates.sort(function(cert1, cert2) {
+        return (cert1.notBefore - cert2.notBefore);
+      });
+
+      if (! validCertificates) {
+        // TODO: Be more robust here, though it's not clear how.
```

Comment:
I partly like the idea of choosing a bogus certificate so it's more obvious the problem is with certificates, but I don't like that this would result in users seeing a "OH NO YOUR BROWSER DETECTED SOMETHING TERRIBLE" message.

I think since re-eying only takes ~30 seconds, what I'd want to do is here is:

* Log a message indicating what's up, and
* Don't listen on port 443 yet, and
* Trigger re-keying, and
* When re-keying succeeds, restart the Sandstorm shell by doing `process.exit(99)` and allowing the auto restart process to restart the Sandstorm shell.

I want to restart at the end because the alternative is to provide more and more monkey-patching and proxy object-ing inside this monkey-patched fakeCreateServer() and I don't know that I can confidently do that reliably. It would require not just a mock `createServer` but a mock `HTTP.Server` instance that Meteor will register callbacks on, which seems a bit terrifying.

Possibly someone should tell me to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy instead (so long as we have those in the version of Meteor we're using).

But note that having 0 key files is a pretty uncommon situation -- anyone who uses the install script will have key files generated during install. This is just for people switching HTTPS on after the fact. So I think that a lot of careful engineering effort in an unlikely codepath is not clearly a good idea.